// arangojs/lib/cjs/analyzer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = require("./error");
const ANALYZER_NOT_FOUND = 1202;
class ArangoAnalyzer {
    constructor(connection, name) {
        this.isArangoAnalyzer = true;
        this._connection = connection;
        this.name = name;
    }
    get() {
        return this._connection.request({ path: `/_api/analyzer/${this.name}` }, res => res.body);
    }
    exists() {
        return this.get().then(() => true, err => {
            if (error_1.isArangoError(err) && err.errorNum === ANALYZER_NOT_FOUND) {
                return false;
            }
            throw err;
        });
    }
    create(options) {
        return this._connection.request({
            method: "POST",
            path: "/_api/analyzer",
            body: Object.assign({ name: this.name }, options)
        }, res => res.body);
    }
    drop(force) {
        return this._connection.request({
            method: "DELETE",
            path: `/_api/analyzer/${this.name}`,
            qs: { force }
        }, res => res.body);
    }
}
exports.ArangoAnalyzer = ArangoAnalyzer;
