// arangojs/lib/cjs/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = require("./error");
const TRANSACTION_NOT_FOUND = 10;
class ArangoTransaction {
    constructor(connection, id) {
        this.isArangoTransaction = true;
        this._connection = connection;
        this.id = id;
    }
    exists() {
        return this.get().then(() => true, err => {
            if (error_1.isArangoError(err) && err.errorNum === TRANSACTION_NOT_FOUND) {
                return false;
            }
            throw err;
        });
    }
    get() {
        return this._connection.request({
            path: `/_api/transaction/${this.id}`
        }, res => res.body.result);
    }
    commit() {
        return this._connection.request({
            method: "PUT",
            path: `/_api/transaction/${this.id}`
        }, res => res.body.result);
    }
    abort() {
        return this._connection.request({
            method: "DELETE",
            path: `/_api/transaction/${this.id}`
        }, res => res.body.result);
    }
    run(fn) {
        this._connection.setTransactionId(this.id);
        try {
            return Promise.resolve(fn());
        }
        finally {
            this._connection.clearTransactionId();
        }
    }
}
exports.ArangoTransaction = ArangoTransaction;
